subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -173,27 +192,40 @@ private String getFilename(Path dir) {
 
         @Override
         public FileVisitResult visitFile(Path file, @Nullable BasicFileAttributes attrs) {
-            String name = stringInterner.intern(file.getFileName().toString());
+            String name = intern(file.getFileName().toString());
             if (isAllowed(file, name, false, attrs, builder.getRelativePath())) {
                 if (attrs == null) {
                     throw new FileSnapshottingException(String.format("Cannot read file '%s': not authorized.", file));
                 }
-                if (attrs.isSymbolicLink()) {
-                    // when FileVisitOption.FOLLOW_LINKS, we only get here when link couldn't be followed
-                    throw new FileSnapshottingException(String.format("Could not list contents of '%s'. Couldn't follow symbolic link.", file));
-                }
-                addFileSnapshot(file, name, attrs);
+                FileSystemLocationSnapshot snapshot = getSnapshot(file, name, attrs);
+                builder.visit(snapshot);
             }
             return FileVisitResult.CONTINUE;
         }
 
+        private FileSystemLocationSnapshot getSnapshot(Path file, String name, BasicFileAttributes attrs) {
+            String absolutePath = intern(file.toString());
+            try {
+                HashCode hash = hasher.hash(file.toFile(), attrs.size(), attrs.lastModifiedTime().toMillis());
```

Comment:
Shall we check that `attrs.isRegularFile() == true` before trying to hash it?

Reply:
If you think that will help, I'll do.
My thinking was that exceptions should happen rarely, so I didn't want to burden the happy path.

Reply:
Done, but now broken symlinks are `Unavailable`, instead of throwing an exception. Will continue addressing that in a new PR